### DISCUSSION TITLE
4 bit color (slideshow)

### DISCUSSION THREAD

**Discussion by michaelsarr** [picocomputer/community#30] [Q&A]
![IMG_6108](https://github.com/picocomputer/discussions/assets/73266612/fa69a7e1-2ad0-4686-b078-08864034beaa)
![IMG_6106](https://github.com/picocomputer/discussions/assets/73266612/90bbdfac-9b6d-4ebc-b3f7-25f790e29d13)

![IMG_6107](https://github.com/picocomputer/discussions/assets/73266612/0904422b-a7c7-4cfd-9617-5093fa119e5e)

ive been playing around a bit trying to push an image to the display
the color palette is pretty rough.. been a long time since the 80s i guess.

I'm interested in loading a more recent version of the vga  is it stable?

does this logic for the color look ok?

//------------------------------------------------------------------------------------------

anyhow currently im opening an image file (320x240)  and trying to convert it to 4 bit color with the following
I run this on my laptop to convert image.jpg to image.bin

```
//-------------------------------------------------------------------------------------------
// slide show
#include <iostream>
#include <stdio.h>

extern "C" {
    #define STB_IMAGE_IMPLEMENTATION
    #include "../inc/stb_image.h"
}




int8_t get4bitcolor(int r,int g, int b);

int main(int argc, char *argv[] )
{
   int width, height,n,size;
   unsigned char temp;
   char filename[100];       
   FILE * outfile;
   int counter=1;;
    if (argc <2)
      printf("usage: image_parser filename(s) \n\n");
   while (counter < argc)
   {
    strncpy(filename,argv[counter++],95  );
    unsigned char* image = stbi_load(filename, &width, &height, &n, 4);
                
    printf("opened %s (%dx%d)\n",filename,width,height);   
    filename[strlen(filename)-4]=0;
    strcat(filename,".bin");
    printf("output: %s\n",filename);
    outfile=fopen(filename,"wb");
    if (!outfile)
    {
    return 1;
    }
    size = width * height*4;

    for (int loop=0;loop< size;loop+=8)
    {
    temp = get4bitcolor(image[loop+0],image[loop+1],image[loop+2]);
    temp <<= 4;
    temp |= get4bitcolor(image[loop+4],image[loop+5],image[loop+6]);
    fwrite( &temp,1,1,outfile);
    }
    fclose(outfile);
   }
    return 0;
} 



int8_t get4bitcolor(int r,int g, int b)
{
int8_t temp=0;
if (r >= 86) temp |= 0b0001;
if (g >= 86) temp |= 0b0010;
if (b >= 86) temp |= 0b0100;
if (r >= 171 || g >= 171 || b >= 171 ) temp |= 0b1000;
return temp;

}

```
//-------------------------------------------------------------------------------------------

then on the picocomputer here is the slide show code to load and display it with,

this is just the brot32.c in the examples used as a starting point

i put the file(s) image_name.bin on the thumb drive and put it in the picocomputer

```
//-------------------------------------------------------------------------------------------

#include <stdint.h>
#include <stdio.h>
#include "rp6502.h"


#define WIDTH 320
#define HEIGHT 240 // 180 or 240

static void vmode(uint16_t data)
{
    xreg(data, 0, 1);
}


void main()
{
FILE * input;
int8_t pixel;
int loop;
int cnt=0;
// names of slides
char names[5][20]={"possum.bin","bivvy.bin","anita.bin","brutus.bin","lucy.bin"};
int slides=5;// number of slides
#if (HEIGHT == 180)
   vmode(2);
#else
    vmode(1);
#endif
  
   while (1)
   {
    RIA_ADDR0 = 0;
    RIA_STEP0 = 1;
    if(cnt == slides) 
    cnt = 0;
    input = fopen(names[cnt],"rb");
    while (!feof(input)) 
    {
     fread(&pixel,1,1,input);
     RIA_RW0 = pixel;
     }
     fclose(input);

     cnt++;
   }

}
```

**Comment by brentward** [comment]
Very cool. Thanks for sharing. I also used brot32.c as the starting point for my SNES controller drawing program.

**Comment by rumbledethumps** [comment]
The vga branch has a working readx() which can load the whole image in under 1 second. This bypasses system RAM (like DirectStorage). I imagined readx() as the primary mechanism for loading game levels. Makes it hard to justify having everyone pay for the complexity of paged RAM in the reference design when loading game levels only takes 1-2 seconds.
```
        #include <fcntl.h>
        fd = open(names[cnt], O_RDONLY);
        if (fd < 0)
        {
            puts("file error");
            exit(1);
        }
        readx(0x0000, 0x4B00, fd);
        readx(0x4B00, 0x4B00, fd);
        close(fd);
        if (++cnt == slides)
            cnt = 0;
```

**Reply by michaelsarr** [reply]
I built the vga and Via  UF2 files and got the readx working, thanks
the slow bit by bit load felt more RETRO but speed is good.

**Comment by rumbledethumps** [comment]
![birbs](https://github.com/picocomputer/discussions/assets/16963588/c1abffd7-0aa9-4dee-9449-61190b872692)

**Reply by michaelsarr** [reply]
Pretty cool. Did you load the image directly from an image file? 
 
When I was playing around with this I didn’t know how large the headers were for the different formats so I was using a library to parse the file first and write out a binary file with only the color info for each pixell sequentialy. I was thinking I would add a small header something like uint16 width, uint16 height, uint8 bits per pixel.   Then a load function   LoadImage(constant char * filename,uint16 x,uint16 y).    So you could load images of different sizes and place them at a specific location.

**Comment by rumbledethumps** [comment]
I used python.
```
from PIL import Image

def rp6502_rgb(r,g,b):
    return (((b>>3)<<11)|((g>>3)<<6)|(r>>3))

try:
    with open("src/kodim23_320.bin", "wb") as o:
        with Image.open('src/kodim23_320.png') as im:
            for r,g,b in zip(*[iter(im.getpalette())]*3):
                o.write(rp6502_rgb(r,g,b).to_bytes(2, byteorder='little', signed=False))
            for y in range(0, im.height):
                for x in range(0, im.width):
                    o.write(im.getpixel((x,y)).to_bytes(1, byteorder='little', signed=False))
except OSError:
    print("cannot")
```

**Reply by michaelsarr** [reply]
ok , that looks similar to  what I did. 
 I havent worked out how to get mode 3 up and displaying stuff yet but ill play around with it some more

**Reply by michaelsarr** [reply]
i havent had any luck getting 16 bit color bitmap working
i was playing around compiling the RIA and VGA from the dev branch but i just grabbed the uf2 files posted with v0.2 
and i have the same result

i have code working for 4 and 8 bit color
when I try to set mode 2 for 320x180, with 

1:0:02  set to 3 for 16 bit the 
xreg returns 0 indicating that it worked, it switches to graphics mode 
but when I put data into memory at 0x0000
it would seem that if I follow the same logic that works for 4 and 8 bit color it would work for 16
but it doesnt just a blank screen

**Reply by rumbledethumps** [reply]
What's the size of your mode 3 16bpp image? Does it fit in 64K?

**Reply by michaelsarr** [reply]
i was trying to load it in pieces
readx(0x0000, 0x4B00, fd);
so there should have been plenty of ram to load the first section

i did a for loop to draw a color bar in the first 80 rows as well , works when 
i use 8 bit or 4 bit mode but nothing on 16 bit
 
 ill post a bit of code tomorrow

**Comment by michaelsarr** [comment]
erase();

xram0_struct_set(0xFF00, vga_mode3_config_t, x_wrap, false);
xram0_struct_set(0xFF00, vga_mode3_config_t, y_wrap, false);
xram0_struct_set(0xFF00, vga_mode3_config_t, x_pos_px, 0);
xram0_struct_set(0xFF00, vga_mode3_config_t, y_pos_px, 0);
xram0_struct_set(0xFF00, vga_mode3_config_t, width_px, 320);
xram0_struct_set(0xFF00, vga_mode3_config_t, height_px, 180);
xram0_struct_set(0xFF00, vga_mode3_config_t, xram_data_ptr, 0x0000);
xram0_struct_set(0xFF00, vga_mode3_config_t, xram_palette_ptr, 0xFFFF);


#define CANVAS 1
#define MODE 3
// this works for 1  and 2  but I only have the text for value 3
#define BPP 3   
#define CONFIG 0xFF00
#define PLANE 0
#define START_SL 0
#define END_SL 180
printf("BPP:%d\n",BPP);
result = xreg(1, 0, 0, CANVAS);
printf("xreg:%d\n",result);
result = xreg(1, 0, 1, MODE, BPP , CONFIG,PLANE ,START_SL,END_SL);
printf("xreg:%d\n",result);
result = xreg(1, 0, 1, 0, 1);// console text on 1
printf("xreg:%d\n",result);


    input  = open("Color_Bars_8_320x240.bin", O_RDONLY);
    result = read_xram(0x0000, 0x4B00, input);
    printf("loaded %d bytes\n",result);
BPP 2
![IMG_6230](https://github.com/picocomputer/discussions/assets/73266612/a521c2b6-5a5d-4f4a-aed4-5e8cec9065c1)

BPP 1
![IMG_6231](https://github.com/picocomputer/discussions/assets/73266612/e0d50895-7198-40dc-a9c0-b5bcf3f2a29a)

BPP3
![IMG_6232](https://github.com/picocomputer/discussions/assets/73266612/b21a2ae1-2b5c-48fc-bf7c-2da49fd34904)

im using uf2 files from the rp6502 release v0.2

not sure what im doing wrong but if you have a hint please share

$1:0:02.   to 1     for 4 bit color. works 
$1:0:02.   to 2     for 8 bit color. works 
$1:0:02.   to 3    for 16 bit color does not seem to work, the xreg call returns 0
